README.md
A collection of [LPEG](http://www.inf.puc-rio.br/~roberto/lpeg/lpeg.html) patterns

## Use cases

  - Strict validation of user input
  - Searching free-form input


## Modules

### `core`

A small module implementing commonly used rules from [RFC-5234 appendix B.1](https://tools.ietf.org/html/rfc5234#appendix-B.1)

  - `ALPHA` (pattern)
  - `BIT` (pattern)
  - `CHAR` (pattern)
  - `CR` (pattern)
  - `CRLF` (pattern)
  - `CTL` (pattern)
  - `DIGIT` (pattern)
  - `DQUOTE` (pattern)
  - `HEXDIG` (pattern)
  - `HTAB` (pattern)
  - `LF` (pattern)
  - `LWSP` (pattern)
  - `OCTET` (pattern)
  - `SP` (pattern)
  - `VCHAR` (pattern)
  - `WSP` (pattern)


### `IPv4`

  - `IPv4address` (pattern): parses an IPv4 address in dotted decimal notation. on success, returns addresses as an IPv4 object
  - `IPv4_methods` (table):
      - `unpack` (function): the IPv4 address as a series of 4 8 bit numbers
      - `binary` (function): the IPv4 address as a 4 byte binary string
  - `IPv4_mt` (table): metatable given to IPv4 objects
      - `__index` (table): `IPv4_methods`
      - `__tostring` (function): returns the IPv4 address in dotted decimal notation

IPv4 "dotted decimal notation" in this document refers to "strict" form (see [RFC-6943 section 3.1.1](https://tools.ietf.org/html/rfc6943#section-3.1.1)) unless otherwise noted.


### `IPv6`

  - `IPv6address` (pattern): parses an IPv6 address
  - `IPv6addrz` (pattern): parses an IPv6 address with optional "ZoneID" (see [RFC-6874](https://tools.ietf.org/html/rfc6874))
  - `IPv6_methods` (table): methods available on IPv6 objects
      - `unpack` (function): the IPv6 address as a series of 8 16bit numbers, optionally followed by zoneid
      - `binary` (function): the IPv6 address as a 16 byte binary string
      - `setzoneid` (function): set the zoneid of this IPv6 address
  - `IPv6_mt` (table): metatable given to IPv6 objects
      - `__tostring` (function): will return the IPv6 address as a valid IPv6 string


### `uri`

Parses URIs as described in [RFC-3986](https://tools.ietf.org/html/rfc3986).

  - `uri` (pattern): on success, returns a table with fields: (similar to [luasocket](http://w3.impa.br/~diego/software/luasocket/url.html#parse))
      - `scheme`
      - `userinfo`
      - `host`
      - `port`
      - `path`
      - `query`
      - `fragment`
  - `absolute_uri` (pattern): similar to `uri`, but does not permit fragments
  - `uri_reference` (pattern): similar to `uri`, but permits relative URIs
  - `relative_part` (pattern): matches a relative uri not including query and fragment; data is held in named group captures `"userinfo"`, `"host"`, `"port"`, `"path"`
  - `scheme` (pattern): matches the scheme portion of a URI
  - `userinfo` (pattern): matches the userinfo portion of a URI
  - `host` (pattern): matches the host portion of a URI
  - `IP_literal` (pattern): matches an IP based host portion of a URI. Capture is an [IPv4](#IPv4), [IPv6](#IPv6) or IPvFuture object
  - `port` (pattern): matches the port portion of a URI
  - `authority` (pattern): matches the authority portion of a URI; data is held in named group captures of `"userinfo"`, `"host"`, `"port"`
  - `path` (pattern): matches the path portion of a URI. Captures `nil` for the empty path.
  - `segment` (pattern): matches a path segment (a piece of a path without a `/`)
  - `query` (pattern): matches the query portion of a URI
  - `fragment` (pattern): matches the fragment portion of a URI
  - `sane_uri` (pattern): a variant that shouldn't match things that people would not normally consider URIs.
    e.g. uris without a hostname
  - `sane_host` (pattern): a variant that shouldn't match things that people would not normally consider valid hosts.
  - `sane_authority` (pattern): a variant that shouldn't match things that people would not normally consider valid hosts.
  - `pct_encoded` (pattern): matches a percent encoded octet, produces a capture of the normalised form.
  - `sub_delims` (pattern): the set of subcomponent delimeters


### `email`

  - `mailbox` (pattern): the mailbox format: matches either `name_addr` or an addr-spec.
  - `name_addr` (pattern): the name and address format i.e. `Display Name<email@example.com>`
    Has captures of the local_part and the domain. Captures the display name in the named capture `"display"`
  - `email` (pattern): also known as an "addr-spec"; follows [RFC-5322 section 3.4.1](http://tools.ietf.org/html/rfc5322#section-3.4.1) with the internationalisation extensions from [RFC-6532 section 3.1](http://tools.ietf.org/html/rfc6532#section-3.1)
    Has captures of the local_part and the domain
    Be careful trying to reconstruct the email address from the captures; you may need escaping
  - `local_part` (pattern): the bit before the `@` in an email address
  - `domain` (pattern): the bit after the `@` in an email address
  - `email_nocfws` (pattern): a variant that doesn't allow for comments or folding whitespace
  - `local_part_nocfws` (pattern): the bit before the `@` in an email address; no comments or folding whitespace allowed.
  - `domain_nocfws` (pattern):  the bit after the `@` in an email address; no comments or folding whitespace allowed.


### `http`

These patterns should be considered to have non stable APIs.

#### [RFC 4918](https://tools.ietf.org/html/rfc4918)

  - `DAV` (pattern)
  - `Depth` (pattern)
  - `Destination` (pattern)
  - `If` (pattern)
  - `Lock_Token` (pattern)
  - `Overwrite` (pattern)
  - `TimeOut` (pattern)


#### [RFC 5023](https://tools.ietf.org/html/rfc5023)

  - `SLUG` (pattern)


#### [RFC 5323](https://tools.ietf.org/html/rfc5323)

  - `DASL` (pattern)


#### [RFC 5789](https://tools.ietf.org/html/rfc5789)

  - `Accept_Patch` (pattern)


#### [RFC 5988](https://tools.ietf.org/html/rfc5988)

  - `Link` (pattern)


#### [RFC 6265](https://tools.ietf.org/html/rfc6265)

  - `Set_Cookie` (pattern)
  - `Cookie` (pattern)


#### [RFC 6266](https://tools.ietf.org/html/rfc6266)

  - `Content_Disposition` (pattern)


#### [RFC 6454](https://tools.ietf.org/html/rfc6454)

  - `Origin` (pattern)


#### [RFC 6455](https://tools.ietf.org/html/rfc6455)

  - `Sec_WebSocket_Accept` (pattern)
  - `Sec_WebSocket_Key` (pattern)
  - `Sec_WebSocket_Extensions` (pattern)
  - `Sec_WebSocket_Protocol_Client` (pattern)
  - `Sec_WebSocket_Protocol_Server` (pattern)
  - `Sec_WebSocket_Version_Client` (pattern)
  - `Sec_WebSocket_Version_Server` (pattern)


#### [RFC 6638](https://tools.ietf.org/html/rfc6638)

  - `Schedule_Reply` (pattern)
  - `Schedule_Tag` (pattern)
  - `If_Schedule_Tag_Match` (pattern)


#### [RFC 6797](https://tools.ietf.org/html/rfc6797)

  - `Strict_Transport_Security` (pattern)


#### [RFC 7034](https://tools.ietf.org/html/rfc7034)

  - `X_Frame_Options` (pattern)


#### [RFC 7089](https://tools.ietf.org/html/rfc7089)

  - `Accept_Datetime` (pattern)
  - `Memento_Datetime` (pattern)


#### [RFC 7230](https://tools.ietf.org/html/rfc7230)

  - `request_line` (pattern)
  - `field_name` (pattern)
  - `field_value` (pattern)
  - `header_field` (pattern)
  - `OWS` (pattern)
  - `RWS` (pattern)
  - `BWS` (pattern)
  - `token` (pattern)
  - `qdtext` (pattern)
  - `quoted_string` (pattern)
  - `comment` (pattern)
  - `Content_Length` (pattern)
  - `Transfer_Encoding` (pattern)
  - `chunk_ext` (pattern)
  - `TE` (pattern)
  - `Trailer` (pattern)
  - `request_target` (pattern)
  - `Host` (pattern)
  - `Via` (pattern): captures are a list of tables with fields `.protocol`, `.by` and `.comment`
  - `Connection` (pattern)
  - `Upgrade` (pattern): captures are a list of strings containing *protocol* or *protocol/version*


#### [RFC 7231](https://tools.ietf.org/html/rfc7231)

  - `IMF_fixdate` (pattern)
  - `Content_Encoding` (pattern)
  - `Content_Type` (pattern)
  - `Content_Language` (pattern)
  - `Content_Location` (pattern)
  - `Expect` (pattern)
  - `Max_Forwards` (pattern)
  - `Accept` (pattern)
  - `Accept_Charset` (pattern)
  - `Accept_Encoding` (pattern)
  - `Accept_Language` (pattern)
  - `From` (pattern)
  - `Referer` (pattern)
  - `User_Agent` (pattern)
  - `Date` (pattern): capture is a table in the same format as used by [`os.time`](http://www.lua.org/manual/5.3/manual.html#pdf-os.time)
  - `Location` (pattern)
  - `Retry_After` (pattern): capture is either a table describing an absolute time in the same format as used by [`os.time`](http://www.lua.org/manual/5.3/manual.html#pdf-os.time), or a relative time as a number of seconds
  - `Vary` (pattern)
  - `Allow` (pattern)
  - `Server` (pattern)


#### [RFC 7232](https://tools.ietf.org/html/rfc7232)

  - `Last_Modified` (pattern): capture is a table in the same format as used by [`os.time`](http://www.lua.org/manual/5.3/manual.html#pdf-os.time)
  - `ETag` (pattern)
  - `If_Match` (pattern)
  - `If_None_Match` (pattern)
  - `If_Modified_Since` (pattern): capture is a table in the same format as used by [`os.time`](http://www.lua.org/manual/5.3/manual.html#pdf-os.time)
  - `If_Unmodified_Since` (pattern): capture is a table in the same format as used by [`os.time`](http://www.lua.org/manual/5.3/manual.html#pdf-os.time)


#### [RFC 7233](https://tools.ietf.org/html/rfc7233)

  - `Accept_Ranges` (pattern)
  - `Range` (pattern)
  - `If_Range` (pattern): capture is either an `entity_tag` or a table in the same format as used by [`os.time`](http://www.lua.org/manual/5.3/manual.html#pdf-os.time)
  - `Content_Range` (pattern)


#### [RFC 7234](https://tools.ietf.org/html/rfc7234)

  - `Age` (pattern)
  - `Cache_Control` (pattern): captures are grouped into key/value pairs (where a directive with no value has a value of `true`)
  - `Expires` (pattern): capture is a table in the same format as used by [`os.time`](http://www.lua.org/manual/5.3/manual.html#pdf-os.time)
  - `Pragma` (pattern)
  - `Warning` (pattern)


#### [RFC 7235](https://tools.ietf.org/html/rfc7235)

  - `WWW_Authenticate` (pattern)
  - `Authorization` (pattern)
  - `Proxy_Authenticate` (pattern)
  - `Proxy_Authorization` (pattern)


#### [RFC 7239](https://tools.ietf.org/html/rfc7239)

  - `Forwarded` (pattern)


#### [RFC 7469](https://tools.ietf.org/html/rfc7469)

  - `Public_Key_Pins` (pattern)
  - `Public_Key_Pins_Report_Only` (pattern)


#### [RFC 7486](https://tools.ietf.org/html/rfc7486)

  - `Hobareg` (pattern)


#### [RFC 7615](https://tools.ietf.org/html/rfc7615)

  - `Authentication_Info` (pattern)
  - `Proxy_Authentication_Info` (pattern)


#### [RFC 7639](https://tools.ietf.org/html/rfc7639)

  - `ALPN` (pattern)


#### [RFC 7809](https://tools.ietf.org/html/rfc7809)

  - `CalDAV_Timezones` (pattern)


#### [RFC 7838](https://tools.ietf.org/html/rfc7838)

  - `Alt_Svc` (pattern)
  - `Alt_Used` (pattern)


#### [Expect-CT Extension for HTTP](https://tools.ietf.org/html/draft-ietf-httpbis-expect-ct-06)

  - `Expect_CT` (pattern)


#### [Referrer-Policy header](https://www.w3.org/TR/referrer-policy/#referrer-policy-header)

  - `Referrer_Policy` (pattern)


### `phone`

  - `phone` (pattern): includes detailed checking for:
      - USA phone numbers using the [NANP](https://en.wikipedia.org/wiki/North_American_Numbering_Plan)


### `language`

Patterns for definitions from [RFC-4646 Section 2.1](https://tools.ietf.org/html/rfc4646#section-2.1)

  - `langtag` (pattern): Capture is a table with the language tag decomposed into components:
      - `language`
      - `extlang` (optional)
      - `script` (optional)
      - `region` (optional)
      - `variant` (optional): an array
      - `extension` (optional): a dictionary from singleton to value
      - `privateuse` (optional): an array
  - `privateuse` (pattern): captures an array
  - `Language_Tag` (pattern): captures the whole language tag


### `utf8`

Patterns for matching utf8 sequences from [RFC 3629 Section 4](https://tools.ietf.org/html/rfc3629#section-4)

  - `UTF8_1` (pattern): matches a 1-byte utf8 sequence
  - `UTF8_2` (pattern): matches a 2-byte utf8 sequence
  - `UTF8_3` (pattern): matches a 3-byte utf8 sequence
  - `UTF8_4` (pattern): matches a 4-byte utf8 sequence
  - `UTF8_char` (pattern): matches a single utf8 sequence
  - `UTF8_octets` (pattern): matches multiple utf8 sequences in a row


NEWS
UNRELEASED

  - http: split into multiple modules
  - http.alpn: verify that protocol_id meets unique encoding criteria
  - utf8: Add new utf8 module
  - email: Allow internationalised email addresses (#15)


0.5 - 2018-07-15

  - http: Cache-Control directives are case-normalised and grouped into pairs
  - http: Strict_Transport_Security now returns a table and doesn't match on duplicates
  - http: Public_Key_Pins capture format and validation
  - http: New Expect_CT and Referrer_Policy patterns


0.4 - 2016-11-23

  - Reduce memory usage by refactoring IPv6 pattern
    Also fixes lpeg 0.10 compatability
  - http: case-normalise cookie attribute names
  - http: fix captures of Via header
  - http: fixes some whitespace rules
  - uri: fix missing case-normalisation for percent encoded characters in hostnames
  - uri: export IP_literal and sub_delims patterns


0.3 - 2016-08-21

  - New http module
    Includes parsers for almost every HTTP header.
    Expect this API to be unstable, it's a significant amount of new code
  - New language module that parses language codes e.g. "zh-Hans-CN"
  - New email.mailbox pattern (name + email like: "Bob <bob@smith.com>")
  - New uri.absolute_uri pattern that does not allow fragments
  - Expose some previously internal uri patterns
  - Fix: Don't percent decode in URIs when it could introduce ambiguity (thanks @torhve)


0.2 - 2015-12-14

  - Fixed parsing of IPv6 addresses (thanks Sean Conner)
  - IPv6 zone support
  - Stricter uri matching (scheme is now compulsory)
  - "reference" (i.e. relative) URI matching


0.1 - 2015-01-29

  - First release


lpeg_patterns/http.lua
-- HTTP related patterns

local _M = {}

-- RFC 7230
local http_core = require "lpeg_patterns.http.core"
_M.OWS = http_core.OWS
_M.RWS = http_core.RWS
_M.BWS = http_core.BWS

_M.chunk_ext = http_core.chunk_ext
_M.comment = http_core.comment
_M.field_name = http_core.field_name
_M.field_value = http_core.field_value
_M.header_field = http_core.header_field
_M.qdtext = http_core.qdtext
_M.quoted_string = http_core.quoted_string
_M.request_line = http_core.request_line
_M.request_target = http_core.request_target
_M.token = http_core.token

_M.Connection = http_core.Connection
_M.Content_Length = http_core.Content_Length
_M.Host = http_core.Host
_M.TE = http_core.TE
_M.Trailer = http_core.Trailer
_M.Transfer_Encoding = http_core.Transfer_Encoding
_M.Upgrade = http_core.Upgrade
_M.Via = http_core.Via

-- RFC 7231
local http_semantics = require "lpeg_patterns.http.semantics"

_M.IMF_fixdate = http_semantics.IMF_fixdate

_M.Accept = http_semantics.Accept
_M.Accept_Charset = http_semantics.Accept_Charset
_M.Accept_Encoding = http_semantics.Accept_Encoding
_M.Accept_Language = http_semantics.Accept_Language
_M.Allow = http_semantics.Allow
_M.Content_Encoding = http_semantics.Content_Encoding
_M.Content_Language = http_semantics.Content_Language
_M.Content_Location = http_semantics.Content_Location
_M.Content_Type = http_semantics.Content_Type
_M.Date = http_semantics.Date
_M.Expect = http_semantics.Expect
_M.From = http_semantics.From
_M.Location = http_semantics.Location
_M.Max_Forwards = http_semantics.Max_Forwards
_M.Referer = http_semantics.Referer
_M.Retry_After = http_semantics.Retry_After
_M.Server = http_semantics.Server
_M.User_Agent = http_semantics.User_Agent
_M.Vary = http_semantics.Vary

-- RFC 7232
local http_conditional = require "lpeg_patterns.http.conditional"
_M.ETag = http_conditional.ETag
_M.If_Match = http_conditional.If_Match
_M.If_Modified_Since = http_conditional.If_Modified_Since
_M.If_None_Match = http_conditional.If_None_Match
_M.If_Unmodified_Since = http_conditional.If_Unmodified_Since
_M.Last_Modified = http_conditional.Last_Modified

-- RFC 7233
local http_range = require "lpeg_patterns.http.range"
_M.Accept_Ranges = http_range.Accept_Ranges
_M.Range = http_range.Range
_M.If_Range = http_range.If_Range
_M.Content_Range = http_range.Content_Range

-- RFC 7234
local http_caching = require "lpeg_patterns.http.caching"
_M.Age = http_caching.Age
_M.Cache_Control = http_caching.Cache_Control
_M.Expires = http_caching.Expires
_M.Pragma = http_caching.Pragma
_M.Warning = http_caching.Warning

-- RFC 7235
local http_authentication = require "lpeg_patterns.http.authentication"
_M.WWW_Authenticate = http_authentication.WWW_Authenticate
_M.Authorization = http_authentication.Authorization
_M.Proxy_Authenticate = http_authentication.Proxy_Authenticate
_M.Proxy_Authorization = http_authentication.Proxy_Authorization

-- WebDav
local http_webdav = require "lpeg_patterns.http.webdav"
_M.CalDAV_Timezones = http_webdav.CalDAV_Timezones
_M.DASL = http_webdav.DASL
_M.DAV = http_webdav.DAV
_M.Depth = http_webdav.Depth
_M.Destination = http_webdav.Destination
_M.If = http_webdav.If
_M.If_Schedule_Tag_Match = http_webdav.If_Schedule_Tag_Match
_M.Lock_Token = http_webdav.Lock_Token
_M.Overwrite = http_webdav.Overwrite
_M.Schedule_Reply = http_webdav.Schedule_Reply
_M.Schedule_Tag = http_webdav.Schedule_Tag
_M.TimeOut = http_webdav.TimeOut

-- RFC 5023
_M.SLUG = require "lpeg_patterns.http.slug".SLUG

-- RFC 5789
_M.Accept_Patch = http_core.comma_sep_trim(http_semantics.media_type, 1)

-- RFC 5988
_M.Link = require "lpeg_patterns.http.link".Link

-- RFC 6265
local http_cookie = require "lpeg_patterns.http.cookie"
_M.Cookie = http_cookie.Cookie
_M.Set_Cookie = http_cookie.Set_Cookie

-- RFC 6266
_M.Content_Disposition = require "lpeg_patterns.http.disposition".Content_Disposition

-- RFC 6454
_M.Origin = require "lpeg_patterns.http.origin".Origin

-- RFC 6455
local http_websocket = require "lpeg_patterns.http.websocket"
_M.Sec_WebSocket_Accept = http_websocket.Sec_WebSocket_Accept
_M.Sec_WebSocket_Key = http_websocket.Sec_WebSocket_Key
_M.Sec_WebSocket_Extensions = http_websocket.Sec_WebSocket_Extensions
_M.Sec_WebSocket_Protocol_Client = http_websocket.Sec_WebSocket_Protocol_Client
_M.Sec_WebSocket_Protocol_Server = http_websocket.Sec_WebSocket_Protocol_Server
_M.Sec_WebSocket_Version_Client = http_websocket.Sec_WebSocket_Version_Client
_M.Sec_WebSocket_Version_Server = http_websocket.Sec_WebSocket_Version_Server

-- RFC 6797
_M.Strict_Transport_Security = require "lpeg_patterns.http.sts".Strict_Transport_Security

-- RFC 7034
_M.X_Frame_Options = require "lpeg_patterns.http.frameoptions".X_Frame_Options

-- RFC 7089
_M.Accept_Datetime = http_semantics.IMF_fixdate
_M.Memento_Datetime = http_semantics.IMF_fixdate

-- RFC 7239
_M.Forwarded = require "lpeg_patterns.http.forwarded".Forwarded

-- RFC 7469
local http_pkp = require "lpeg_patterns.http.pkp"
_M.Public_Key_Pins = http_pkp.Public_Key_Pins
_M.Public_Key_Pins_Report_Only = http_pkp.Public_Key_Pins_Report_Only

-- RFC 7486
_M.Hobareg = require "lpeg_patterns.http.hoba".Hobareg

-- RFC 7615
_M.Authentication_Info = http_authentication.Authentication_Info
_M.Proxy_Authentication_Info = http_authentication.Proxy_Authentication_Info

-- RFC 7639
_M.ALPN = require "lpeg_patterns.http.alpn".ALPN

-- RFC 7838
local http_alternate = require "lpeg_patterns.http.alternate"
_M.Alt_Svc = http_alternate.Alt_Svc
_M.Alt_Used = http_alternate.Alt_Used

-- https://tools.ietf.org/html/draft-ietf-httpbis-expect-ct-06#section-2.1
_M.Expect_CT = require "lpeg_patterns.http.expect_ct".Expect_CT

-- https://www.w3.org/TR/referrer-policy/#referrer-policy-header
_M.Referrer_Policy = require "lpeg_patterns.http.referrer_policy".Referrer_Policy

return _M
